<center>
<img src="../../img/ods_stickers.jpg">
    
## [mlcourse.ai](https://mlcourse.ai) - Open Machine Learning Course

Author: [Yury Kashnitsky](https://www.linkedin.com/in/festline/). Translated and edited by [Sergey Isaev](https://www.linkedin.com/in/isvforall/), [Artem Trunov](https://www.linkedin.com/in/datamove/), [Anastasia Manokhina](https://www.linkedin.com/in/anastasiamanokhina/), and [Yuanyuan Pao](https://www.linkedin.com/in/yuanyuanpao/). All content is distributed under the [Creative Commons CC BY-NC-SA 4.0](https://creativecommons.org/licenses/by-nc-sa/4.0/) license.

# <center> Assignment #1 (demo)
## <center>  Exploratory data analysis with Pandas


**Same assignment as a [Kaggle Kernel](https://www.kaggle.com/kashnitsky/a1-demo-pandas-and-uci-adult-dataset) + [solution](https://www.kaggle.com/kashnitsky/a1-demo-pandas-and-uci-adult-dataset-solution).**

**In this task you should use Pandas to answer a few questions about the [Adult](https://archive.ics.uci.edu/ml/datasets/Adult) dataset. (You don't have to download the data – it's already  in the repository). Choose the answers in the [web-form](https://docs.google.com/forms/d/1uY7MpI2trKx6FLWZte0uVh3ULV4Cm_tDud0VDFGCOKg).**

Unique values of all features (for more information, please see the links above):
- `age`: continuous.
- `workclass`: Private, Self-emp-not-inc, Self-emp-inc, Federal-gov, Local-gov, State-gov, Without-pay, Never-worked.
- `fnlwgt`: continuous.
- `education`: Bachelors, Some-college, 11th, HS-grad, Prof-school, Assoc-acdm, Assoc-voc, 9th, 7th-8th, 12th, Masters, 1st-4th, 10th, Doctorate, 5th-6th, Preschool.
- `education-num`: continuous.
- `marital-status`: Married-civ-spouse, Divorced, Never-married, Separated, Widowed, Married-spouse-absent, Married-AF-spouse.
- `occupation`: Tech-support, Craft-repair, Other-service, Sales, Exec-managerial, Prof-specialty, Handlers-cleaners, Machine-op-inspct, Adm-clerical, Farming-fishing, Transport-moving, Priv-house-serv, Protective-serv, Armed-Forces.
- `relationship`: Wife, Own-child, Husband, Not-in-family, Other-relative, Unmarried.
- `race`: White, Asian-Pac-Islander, Amer-Indian-Eskimo, Other, Black.
- `sex`: Female, Male.
- `capital-gain`: continuous.
- `capital-loss`: continuous.
- `hours-per-week`: continuous.
- `native-country`: United-States, Cambodia, England, Puerto-Rico, Canada, Germany, Outlying-US(Guam-USVI-etc), India, Japan, Greece, South, China, Cuba, Iran, Honduras, Philippines, Italy, Poland, Jamaica, Vietnam, Mexico, Portugal, Ireland, France, Dominican-Republic, Laos, Ecuador, Taiwan, Haiti, Columbia, Hungary, Guatemala, Nicaragua, Scotland, Thailand, Yugoslavia, El-Salvador, Trinadad&Tobago, Peru, Hong, Holand-Netherlands.   
- `salary`: >50K,<=50K

In [1]:
import numpy as np
import pandas as pd

pd.set_option("display.max.columns", 100)
# to draw pictures in jupyter notebook
%matplotlib inline
# we don't like warnings
# you can comment the following 2 lines if you'd like to
import warnings

import matplotlib.pyplot as plt
import seaborn as sns

warnings.filterwarnings("ignore")

In [4]:
data = pd.read_csv("../data/adult.data.csv")
data.head()

age         workclass  fnlwgt  education  education-num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   

       marital-status         occupation   relationship   race     sex  \
0       Never-married       Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female   

   capital-gain  capital-loss  hours-per-week native-country salary  
0          2174             0              40  United-States  <=50K  
1             0             0              13  United-States  <=50K  
2             0             0              40  United-States  <=50K  
3             0             0              40  United-States  <=50K  
4             0             0              40           Cuba  <=50K

**1. How many men and women (*sex* feature) are represented in this dataset?** 

In [ ]:
# You code here

In [7]:
data['sex'].value_counts()

Male      21790
Female    10771
Name: sex, dtype: int64

**Ответ:** данные содержат записи о 21790 мужчинах и о 10771 женщине.

**2. What is the average age (*age* feature) of women?**

In [ ]:
# You code here

In [9]:
data[data['sex'] == 'Female']['age'].mean()

36.85823043357163

**Ответ:** средний возраст женщин из данной выборки 36 лет.

**3. What is the percentage of German citizens (*native-country* feature)?**

In [ ]:
# You code here

In [25]:
(data['native-country'] == 'Germany').sum() / data.shape[0] * 100

0.42074874850281013

**Ответ:** 0.42% записей, отражают данные о гражданах Германии.

**4-5. What are mean value and standard deviation of the age of those who recieve more than 50K per year (*salary* feature) and those who receive less than 50K per year?**

In [ ]:
# You code here

In [27]:
data.groupby('salary')['age'].describe()

count       mean        std   min   25%   50%   75%   max
salary                                                             
<=50K   24720.0  36.783738  14.020088  17.0  25.0  34.0  46.0  90.0
>50K     7841.0  44.249841  10.519028  19.0  36.0  44.0  51.0  90.0

**Ответ:** Для получающих зарплату >50K  среднее значение - 44.25, стандартное отклонение - 10.52.  
           Для получающих зарплату <=50K  среднее значение - 36.78, стандартное отклонение - 14.02.

**6. Is it true that people who earn more than 50K have at least high school education? (*education – Bachelors, Prof-school, Assoc-acdm, Assoc-voc, Masters* or *Doctorate* feature)**

In [ ]:
# You code here

In [37]:
data[data["salary"] == ">50K"]["education"].unique() 

array(['HS-grad', 'Masters', 'Bachelors', 'Some-college', 'Assoc-voc',
       'Doctorate', 'Prof-school', 'Assoc-acdm', '7th-8th', '12th',
       '10th', '11th', '9th', '5th-6th', '1st-4th'], dtype=object)

**Ответ:** Нет, утверждение о том что люди зарабатывающие >50K имеют минимум среднее образование неверно.

**7. Display age statistics for each race (*race* feature) and each gender (*sex* feature). Use *groupby()* and *describe()*. Find the maximum age of men of *Amer-Indian-Eskimo* race.**

In [ ]:
# You code here

In [46]:
data.groupby(['race','sex'])['age'].describe()

count       mean        std   min   25%   50%  \
race               sex                                                       
Amer-Indian-Eskimo Female    119.0  37.117647  13.114991  17.0  27.0  36.0   
                   Male      192.0  37.208333  12.049563  17.0  28.0  35.0   
Asian-Pac-Islander Female    346.0  35.089595  12.300845  17.0  25.0  33.0   
                   Male      693.0  39.073593  12.883944  18.0  29.0  37.0   
Black              Female   1555.0  37.854019  12.637197  17.0  28.0  37.0   
                   Male     1569.0  37.682600  12.882612  17.0  27.0  36.0   
Other              Female    109.0  31.678899  11.631599  17.0  23.0  29.0   
                   Male      162.0  34.654321  11.355531  17.0  26.0  32.0   
White              Female   8642.0  36.811618  14.329093  17.0  25.0  35.0   
                   Male    19174.0  39.652498  13.436029  17.0  29.0  38.0   

                             75%   max  
race               sex                  
Amer-Indian-Eskimo Female  46.00  80.0  
                   Male    45.00  82.0  
Asian-Pac-Islander Female  43.75  75.0  
                   Male    46.00  90.0  
Black              Female  46.00  90.0  
                   Male    46.00  90.0  
Other              Female  39.00  74.0  
                   Male    42.00  77.0  
White              Female  46.00  90.0  
                   Male    49.00  90.0

**Ответ:** максимальный возраст для мужчин принадлежащих к Amer-Indian-Eskimo рассе = 82 годам.

**8. Among whom is the proportion of those who earn a lot (>50K) greater: married or single men (*marital-status* feature)? Consider as married those who have a *marital-status* starting with *Married* (Married-civ-spouse, Married-spouse-absent or Married-AF-spouse), the rest are considered bachelors.**

In [ ]:
# You code here

In [56]:
male_ = data[(data['salary'] == '>50K') & (data['sex'] == 'Male')]['marital-status']

count = male_.str.startswith('Married').sum()
male_.shape[0]
print(f'Число женатых мужчин, зарабатывающих >50K - {count}, неженатых - {male_.shape[0] - count}.')

Число женатых мужчин зарабатывающих >50K - 5965, неженатых - 697.


**Ответ:** среди мужчин зарабатывающих >50K больше женатых.

**9. What is the maximum number of hours a person works per week (*hours-per-week* feature)? How many people work such a number of hours, and what is the percentage of those who earn a lot (>50K) among them?**

In [61]:
# You code here

In [74]:
max_hours = data['hours-per-week'].max()
count = data[data['hours-per-week'] == max_hours].shape[0]
print(f'Максимальное кол-во часов отработанных в неделю равно {max_hours}.\
      \nПри этом обнаружено {count} человек, которые отрабатывают данное количество часов')
data[data['hours-per-week'] == max_hours]['salary'].value_counts(normalize=True)

Максимальное кол-во часов отработанных в неделю равно 99.      
При этом обнаружено 85 человек, которые отрабатывают данное количество часов


<=50K    0.705882
>50K     0.294118
Name: salary, dtype: float64

**Ответ:** максимальное кол-во часов отработанных в неделю равно 99'.  
При этом обнаружено 85 человек, которые отрабатывают данное количество часов, при этом доля зарабатывающих >50K равна 29%

**10. Count the average time of work (*hours-per-week*) for those who earn a little and a lot (*salary*) for each country (*native-country*). What will these be for Japan?**

In [ ]:
# You code here

In [90]:
with pd.option_context('display.max_rows', None,
                       'display.max_columns', None,
                       'display.precision', 3,
                       ):
    print(data.groupby(['native-country', 'salary'])['hours-per-week'].mean())

native-country              salary
?                           <=50K     40.165
                            >50K      45.548
Cambodia                    <=50K     41.417
                            >50K      40.000
Canada                      <=50K     37.915
                            >50K      45.641
China                       <=50K     37.382
                            >50K      38.900
Columbia                    <=50K     38.684
                            >50K      50.000
Cuba                        <=50K     37.986
                            >50K      42.440
Dominican-Republic          <=50K     42.338
                            >50K      47.000
Ecuador                     <=50K     38.042
                            >50K      48.750
El-Salvador                 <=50K     36.031
                            >50K      45.000
England                     <=50K     40.483
                            >50K      44.533
France                      <=50K     41.059
                    

In [80]:
pd.crosstab(index=data['native-country'], columns=data['salary'], values=data['hours-per-week'], aggfunc='mean')

salary                          <=50K       >50K
native-country                                  
?                           40.164760  45.547945
Cambodia                    41.416667  40.000000
Canada                      37.914634  45.641026
China                       37.381818  38.900000
Columbia                    38.684211  50.000000
Cuba                        37.985714  42.440000
Dominican-Republic          42.338235  47.000000
Ecuador                     38.041667  48.750000
El-Salvador                 36.030928  45.000000
England                     40.483333  44.533333
France                      41.058824  50.750000
Germany                     39.139785  44.977273
Greece                      41.809524  50.625000
Guatemala                   39.360656  36.666667
Haiti                       36.325000  42.750000
Holand-Netherlands          40.000000        NaN
Honduras                    34.333333  60.000000
Hong                        39.142857  45.000000
Hungary                     31.300000  50.000000
India                       38.233333  46.475000
Iran                        41.440000  47.500000
Ireland                     40.947368  48.000000
Italy                       39.625000  45.400000
Jamaica                     38.239437  41.100000
Japan                       41.000000  47.958333
Laos                        40.375000  40.000000
Mexico                      40.003279  46.575758
Nicaragua                   36.093750  37.500000
Outlying-US(Guam-USVI-etc)  41.857143        NaN
Peru                        35.068966  40.000000
Philippines                 38.065693  43.032787
Poland                      38.166667  39.000000
Portugal                    41.939394  41.500000
Puerto-Rico                 38.470588  39.416667
Scotland                    39.444444  46.666667
South                       40.156250  51.437500
Taiwan                      33.774194  46.800000
Thailand                    42.866667  58.333333
Trinadad&Tobago             37.058824  40.000000
United-States               38.799127  45.505369
Vietnam                     37.193548  39.200000
Yugoslavia                  41.600000  49.500000

**Ответ:** для граждан Японии среднее количество отработанных часов в неделю для зарабатывающих <=50K равно 41.00 , а для >50K равно 47.96